src/notation/view/internal/undoredotoolbarmodel.cpp

```diff
@@ -22,98 +22,96 @@
 
 #include "undoredotoolbarmodel.h"
 
-#include "uicomponents/view/menuitem.h"
+#include "uicomponents/view/toolbaritem.h"
 
 using namespace mu::notation;
 using namespace muse;
 using namespace muse::uicomponents;
 
+static const actions::ActionCode UNDO_ACTION_CODE("undo");
+static const actions::ActionCode REDO_ACTION_CODE("redo");
+
 UndoRedoToolbarModel::UndoRedoToolbarModel(QObject* parent)
-    : QObject(parent), Injectable(iocCtxForQmlObject(this))
+    : AbstractToolBarModel(parent)
 {
 }
 
 void UndoRedoToolbarModel::load()
 {
-    context()->currentNotationChanged().onNotify(this, [this]() {
-        updateItems();
-
-        auto stack = undoStack();
-        if (stack) {
-            stack->stackChanged().onNotify(this, [this]() {
-                updateItems();
-            });
-        }
-    });
-
-    auto stack = undoStack();
-    if (stack) {
-        stack->stackChanged().onNotify(this, [this]() {
-            updateItems();
-        });
+    actions::ActionCodeList itemsCodes = {
+        UNDO_ACTION_CODE,
+        REDO_ACTION_CODE
+    };
+
+    ToolBarItemList items;
+    for (const actions::ActionCode& code : itemsCodes) {
+        ToolBarItem* item = makeItem(code);
+        item->setIsTransparent(true);
+        items << item;
     }
 
-    m_undoItem = new MenuItem(actionsRegister()->action("undo"), this);
-    m_redoItem = new MenuItem(actionsRegister()->action("redo"), this);
+    setItems(items);
 
-    updateItems();
+    AbstractToolBarModel::load();
 
-    // Only to let QML know that `m_undoItem` and `m_redoItem` have been initialised;
-    // changes to their properties will be communicated via those MenuItems' own signals.
-    emit itemsChanged();
-}
+    context()->currentNotationChanged().onNotify(this, [this]() {
+        updateItems();
 
-MenuItem* UndoRedoToolbarModel::undoItem() const
-{
-    return m_undoItem;
-}
+        subsribeOnUndoStackChanges();
+    });
 
-MenuItem* UndoRedoToolbarModel::redoItem() const
-{
-    return m_redoItem;
+    subsribeOnUndoStackChanges();
 }
 
-void UndoRedoToolbarModel::updateItems()
+void UndoRedoToolbarModel::onActionsStateChanges(const muse::actions::ActionCodeList& codes)
 {
     auto stack = undoStack();
-
-    if (m_undoItem) {
-        const TranslatableString undoActionName = stack ? stack->topMostUndoActionName() : TranslatableString();
-        ui::UiActionState state;
-        state.enabled = stack ? stack->canUndo() : false;
-        m_undoItem->setState(state);
-        m_undoItem->setTitle(undoActionName.isEmpty()
-                             ? TranslatableString("action", "Undo")
-                             : TranslatableString("action", "Undo ‘%1’").arg(undoActionName));
+    if (!stack) {
+        return;
```

Comment:
I realised this is wrong: if there is no current notation open, then the actions need to be reset to a disabled state with just `"Undo"` / `"Redo"` as their title. Basically, just need to remove this `if (!stack) { return; }`; then it will work fine.